server/app/org/scalaexercises/exercises/services/ExercisesService.scala

```diff
@@ -110,29 +105,31 @@ sealed trait RuntimeSharedConversions {
       "#0F3057"
     )
 
-    val (_, res) = libraries.foldLeft((autoPalette, Nil: List[RuntimeLibrary])) { (acc, library) ⇒
-      val (colors, librariesAcc) = acc
+    libraries.traverseU { library ⇒
       if (library.color.isEmpty) {
-        val (color, colors0) = colors match {
-          case head :: tail ⇒ Some(head) → tail
-          case Nil          ⇒ None → Nil
+        State[Colors, RuntimeLibrary] { colors ⇒
+          val (color, colors0) = colors match {
+            case head :: tail ⇒ Some(head) → tail
+            case Nil          ⇒ None → Nil
+          }
+          val library0 = DefaultLibrary(
+            owner = library.owner,
+            repository = library.repository,
+            name = library.name,
+            description = library.description,
+            color = color,
+            logoPath = library.logoPath,
+            logoData = library.logoData,
+            sections = library.sections,
+            timestamp = Timestamp.fromDate(new java.util.Date()),
+            buildMetaInfo = library.buildMetaInfo
+          )
+          (colors0, library0)
         }
-        colors0 → (DefaultLibrary(
-          owner = library.owner,
-          repository = library.repository,
-          name = library.name,
-          description = library.description,
-          color = color,
-          logoPath = library.logoPath,
-          logoData = library.logoData,
-          sections = library.sections,
-          timestamp = Timestamp.fromDate(new java.util.Date()),
-          buildMetaInfo = library.buildMetaInfo
-        ) :: librariesAcc)
-      } else
-        colors → (library :: librariesAcc)
-    }
-    res.reverse
+      } else {
+        State.pure[Colors, RuntimeLibrary](library)
+      }
+    }.runA(autoPalette).value
```

Comment:
I prefer the `State[Colors, ?]` with `traverse` over the `foldLeft`, but opinions may vary ?